ncclient/transport/ssh.py

```diff
@@ -371,25 +389,57 @@ def connect(self, host, port=830, timeout=None, unknown_host_cb=default_unknown_
                 sock = socket.fromfd(int(sock_fd), socket.AF_INET, socket.SOCK_STREAM)
             sock.settimeout(timeout)
 
-        t = self._transport = paramiko.Transport(sock)
-        t.set_log_channel(logger.name)
+        self._transport = paramiko.Transport(sock)
+        self._transport.set_log_channel(logger.name)
         if config.get("compression") == 'yes':
-            t.use_compression()
-
+            self._transport.use_compression()
+
+        # Set preferred host keys, to those we possess for the host
+        # Otherwise may have a situation where known_hosts contains a valid key for the host,
+        # but that key is not selected during negotiation
+        if self._host_keys:
+            if port == PORT_SSH_DEFAULT:
+                known_hosts_lookup = host
+            else:
+                known_hosts_lookup = '[%s]:%s' % (host, port)
+            known_host_keys_for_this_host = self._host_keys.lookup(known_hosts_lookup)
+            if known_host_keys_for_this_host:
+                self._transport._preferred_keys = [x.key.get_name() for x in known_host_keys_for_this_host._entries]
+
+        # If we need to connect with a specific hostkey, negotiate for only its type
+        if hostkey:
+            hostkey_obj = None
+            for key_cls in [paramiko.DSSKey, paramiko.Ed25519Key, paramiko.RSAKey, paramiko.ECDSAKey]:
+                try:
+                    hostkey_obj = key_cls(data=base64.b64decode(hostkey))
+                except paramiko.SSHException as e:
+                    # Not a key of this type - try the next
+                    pass
+            if not hostkey_obj:
+                # We've tried all known host key types and haven't found which one to use - bail
+                raise e
+            self._transport._preferred_keys = [hostkey_obj.get_name()]
+    
+        # Connect
         try:
-            t.start_client()
+            self._transport.start_client()
         except paramiko.SSHException:
-            raise SSHError('Negotiation failed')
-
-        # host key verification
-        server_key = t.get_remote_server_key()
-
-        fingerprint = _colonify(hexlify(server_key.get_fingerprint()))
+            # Raise original exception; don't swallow details
+            raise
```

Comment:
1) This is a potentially breaking change for existing users of ncclient who are currently doing "except SSHError" and now need to change to "except paramiko.SSHException". Are we sure this is a desired change?

2) If this is a desired change, there's no point to keeping a try...except block that does nothing but re-raising the exception - just remove the block entirely instead.

Reply:
1. That's actually a good point. What I'll do is revert to SSHError, but put in the detail from paramiko. The reason for this change was that while using the library, I had an error where this failed but the swallowed/rewritten error obfuscated the root cause.

2. Correct there's no point functionally in doing so, only in being more explicit in the type of error we're expecting. Personal thing again; more explicit vs verbosity